### PR TITLE
RFE: Templating Support for Definition File

### LINKED ISSUE #963: RFE: Templating Support for Definition File
Description
-----------
If we can pass values at build time to replace variables' value inside definition file, before actually process definition file to build image, that brings flexibility to how we interact with definition file. I called this feature, "Templateing".

Usually OSS projects distribute not only source code but its binaries and also container images when they release these days. In the CI/CD era, creating binaries is written in the form of scripts that covers wide range of OS and variety of dependencies. That "recipe" can switch OS versions, library versions etc. Widely known examples in HPC are Spack recipes.

If Apptainer definition file support "Templating" feature:

+ Less integration effort for OSS maintainers: They can reuse existing "recipe" scripts when they try to add Apptainer definition file to their release artifact. 
+ It allows users to alter base image and/or version (This is not supported in Apptainer yet.)
+ It allows users to alter script "switch" (This is kind of supported through APPTAINERENV_VARS but this is not support every section in definition file.)
+ It allows users to alter variables at every section in consistent way (This is not supported in Apptainer yet.)

This is inspired by Dockerfile `ARG` and `--build-arg` (`--build-arg-file`) option for build command.

UI Changes
------------------------
Add `--build-arg` and `--build-arg-file` options to `build` command

`--build-arg`

```
apptainer build --build-arg OS_VER=8.7 --build-arg APP_VERSION=2206 app.sif app.def
```

`--build-arg-file`

```
apptainer build --build-arg-file build-args app.sif app.def
```

`build-args` file

```
OS_VER=8.7
APP_VERSION=2206
```

Internal Behavioural Changes
------------------------

When `build-arg` or `build-arg-file` option comes in:
1. process build-args first
2. fill template to create actual definition file
3. process definition file

How definition file looks like:
-------------------------------

```
Bootstrap: docker
From: rockylinux:${OS_VER}

%setup
    touch /${OS_VER}
    touch ${APPTAINER_ROOTFS}/file2

%files
    /script-${OS_VER}.sh
    /dir-${OS_VER} /opt

%environment
    export OS_VER=${OS_VER}

%post
    /script-${OS_VER}.sh ${APP_VER}

%runscript
    echo "Container was created $NOW"
    echo "Arguments received: $*"
    exec echo "$@"

%startscript
    nc -lp ${LISTEN_PORT:-'8080'}

%test
    grep -q NAME=\"Rocky\" /etc/os-release
    if [ $? -eq 0 ]; then
        echo "Container base is Rocky Linux as expected."
    else
        echo "Container base is not Rocky Linux."
        exit 1
    fi

%labels
    Author ${AUTHER:-}
    Version ${VERSION:-}

%help
    This is a ${DEMO:-} for templating definition file
```

I am please to hear any comments and ideas from community here.

### PR DESCRIPTION
**Pull request by JasonYangShadow**
## Description of the Pull Request (PR):

Add template support for build procedure.

### This fixes or addresses the following GitHub issues:

 - Fixes #963

Old PR link:
https://github.com/apptainer/apptainer/pull/1344

#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

### CODE REVIEW

**Comment by JasonYangShadow** [comment]
after rebase I accidentally closed the original PR, so I created this new PR and also linked old PR in the description part. Sorry for the inconvenience.  😢

**Review by DrDaveD — APPROVED** [review]
Looks fine to me, but I'd like @cclerget to also review.

File: pkg/build/types/parser/deffile.go [hunk 1]
```diff
@@ -316,10 +316,11 @@ func populateDefinition(sections map[string]*types.Script, files *[]types.Files,
 	}
 	d.BuildData.Files = *files
 	d.BuildData.Scripts = types.Scripts{
-		Pre:   *sections["pre"],
-		Setup: *sections["setup"],
-		Post:  *sections["post"],
-		Test:  *sections["test"],
+		Default: *sections["default"],
```
Comment (cclerget) [inline pkg/build/types/parser/deffile.go:319]: `default` section is a really generic name, could this be renamed to `arguments` ?

**Review by cclerget — CHANGES_REQUESTED** [review]

File: pkg/build/types/parser/deffile.go (hunk 1 above)
Comment (JasonYangShadow) [reply pkg/build/types/parser/deffile.go]: reflected in this commit
https://github.com/apptainer/apptainer/pull/1356/commits/40513e1ddb9fb0e7a088272640198fa26eb6b3a1

**Review by cclerget — APPROVED** [review]

**Review by dtrudg — COMMENTED** [review]
_Edit_ - apologies, you can delete this. I inadvertandely pasted a review into this PR, instead of the sylabs/singularity PR which is related.

**Pull request merged**